kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/MongoDebeziumTransform.java

```diff
@@ -0,0 +1,334 @@
+                }
+              } else if (field.name().equals(BEFORE_FIELD_NAME)) {
+                if (beforeBson != null) {
+                  mutateBuilderFromBson(newValueSchemaBuilder, beforeBson, BEFORE_FIELD_NAME);
+                }
+              } else {
+                newValueSchemaBuilder.field(field.name(), field.schema());
+              }
+            });
+
+    Schema newValueSchema = newValueSchemaBuilder.build();
+    Struct newValueStruct = new Struct(newValueSchemaBuilder.build());
+
+    newValueSchema
+        .fields()
+        .forEach(
+            field -> {
+              if (field.name().equals(AFTER_FIELD_NAME)) {
+                if (afterBson != null) {
+                  newValueStruct.put(field.name(), fillStructFromBson(field.schema(), afterBson));
+                }
+              } else if (field.name().equals(BEFORE_FIELD_NAME)) {
+                if (beforeBson != null) {
+                  newValueStruct.put(field.name(), fillStructFromBson(field.schema(), beforeBson));
+                }
+              } else {
+                newValueStruct.put(field.name(), oldValue.get(field.name()));
+              }
+            });
+    return record.newRecord(
+        record.topic(),
+        record.kafkaPartition(),
+        newKeySchema,
+        newKeyStruct,
+        newValueSchema,
+        newValueStruct,
+        record.timestamp(),
+        record.headers());
+  }
+
+  private SchemaBuilder mutateBuilderFromBson(
+      SchemaBuilder builder, BsonDocument bson, String fieldName) {
+    SchemaBuilder innerBuilder = SchemaBuilder.struct();
+    Set<Map.Entry<String, BsonValue>> pairs = bson.entrySet();
+    for (Map.Entry<String, BsonValue> pairsForSchema : pairs) {
+      converter.addFieldSchema(pairsForSchema, innerBuilder);
+    }
+    builder.field(fieldName, innerBuilder.optional().build());
+    return builder;
+  }
+
+  private Struct fillStructFromBson(Schema schema, BsonDocument bson) {
+    Struct struct = new Struct(schema);
+    Set<Map.Entry<String, BsonValue>> pairs = bson.entrySet();
+    for (Map.Entry<String, BsonValue> pairsForSchema : pairs) {
+      converter.convertRecord(pairsForSchema, schema, struct);
+    }
+    return struct;
+  }
+
+  private boolean isValidKey(final R record) {
+    if (record.keySchema() == null
+        || record.keySchema().name() == null
+        || !record.keySchema().name().endsWith(RECORD_ENVELOPE_KEY_SCHEMA_NAME_SUFFIX)) {
+      LOG.debug(
+          "Expected Key Schema for transformation, converting to tombstone. Message key: \"{}\"",
+          record.key());
+      return false;
+    }
+    return true;
+  }
+
+  private boolean isValidValue(final R record) {
+    if (record.valueSchema() == null
+        || record.valueSchema().name() == null
+        || !containsEnvelopePrefix(record.valueSchema())) {
+      LOG.debug("Expected Envelope for transformation, converting to tombstone");
+      return false;
+    }
+    return true;
+  }
```

Comment:
This is weird, the log lines about converting to tombstone is nowhere near where we do the actual converting to tombstone. I would do this a little differently to be clearer, something like this: 

```suggestion
  private boolean isValidKey(final SinkRecord record) {
    return record.keySchema() != null
            && record.keySchema().name() != null
            && record.keySchema().name().endsWith(RECORD_ENVELOPE_KEY_SCHEMA_NAME_SUFFIX);
  }

  private boolean isValidValue(final SinkRecord record) {
    return record.valueSchema() != null
            && record.valueSchema().name() != null
            && containsEnvelopePrefix(record.valueSchema());
  }
```

Reply:
I'm on the fence on this is general, but similar logic exists in the existing debezium MongoDebeziumConnector so following it.